src/startup_client.h

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright © 2022 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef FRAME_CRASH_REPORTER
+#define FRAME_CRASH_REPORTER
+
+#include <memory>
+#include <mutex>
+#include <string>
+
+#include <sys/inotify.h>
+
+#include <miral/application.h>
+#include <mir/geometry/rectangles.h>
+
+#include <boost/filesystem.hpp>
+
+#include <ft2build.h>
+#include FT_FREETYPE_H
+
+struct wl_display;
+
+namespace geom = mir::geometry;
+
+struct Pixel
+{
+    uint8_t b;
+    uint8_t g;
+    uint8_t r;
+    uint8_t a;
```

Comment:
I don't think alpha is ever usefully anything but 255
```suggestion
    uint8_t const a = 0xff;
```

Reply:
This `Pixel` is designed to maintain compatibility with the old definition of `Pixel` (`typedef uint8_t Pixel`) which meant a "real" pixel (not a subpixel) had to be accessed as a `Pixel*`.

Assigning the alpha value to a const would break the ability to cast a `uint8_t*` pixel to a `Pixel`. Instead, I think we should set the alpha value to `0xFF` via a default in the `Pixel()` constructor.

Reply:
> ability to cast a uint8_t* pixel to a Pixel

That is undefined behaviour. Best avoided

Reply:
C++ can't change the order of data members, so if you static assert that `sizeof(Pixel) == 4` than essentially it is defined that a `Pixel*` can be coasted to a `uint8_t*`, right? How could it represent a `Pixel` as anything other than the 4 `uint8_t`s in a row?

Reply:
It's a type-based aliasing violation; you are only allowed to dereference memory through a type-compatible pointer or an `unsigned char` pointer.

Now, `uint8_t` is *almost certainly* a typedef to `unsigned char` on every platform we care about, but a compiler assuming any Pointer* accesses don't alias an `uint8_t*` accesses would be technically correct.

Reply:
[this](https://stackoverflow.com/a/2088428) and [this](https://stackoverflow.com/a/60536461) seem to indicate it's fine in at least some cases. It seems that you have to pay for ?? the actual standard, and the versions I found online aren't very searchable so I can't find what exactly it says, but I know the C standard explicitly allows using a pointer to a struct as a pointer to the first member. Is C++ different?

Reply:
[here](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf) is where it is in the C standard, 6.7.2.1 paragraph 13:
> Within a structure object, the non-bit-field members and the units in which bit-fields
reside have addresses that increase in the order in which they are declared. A pointer to a
structure object, suitably con verted, points to its initial member (or if that member is a
bit-field, then to the unit in which it resides), and vice v ersa. There may be unnamed
padding within a structure object, but not at its beginning.

Reply:
Hm. I'm certainly no spec authority, but my reading of that is a guarantee that given
```
struct Foo
{
    int a;
    int b;
};

Foo bar;
```
It's guaranteed that `&bar.a == static_cast<int*>(&bar)`. It doesn't appear to say anything about *access* through that pointer. :woman_shrugging: 

There are then two questions:
1. Is it satisfying the type-based aliasing rules (I'm honestly not sure), and
2. If you want to use the `uint8_t*` as an array, is it guaranteed that `&bar.b == &bar.a + 1` (it is not, compilers are free to add whatever padding they wish).

Reply:
People, the first member of a structure object can be accessed casting the object address/reference to the member type address/reference. This support is motivated by the "tagged variant" idiom.

But that doesn't imply the opposite conversion, from member address to object address is permitted. That is only guaranteed by the standard _if the original address was taken from an actual object_, and only cast via `void*` or `char*` and back. (One reason for this is that the object type may have stricter alignment constraints than the member type - while this doesn't apply here, nobody was motivated to draft standard text to support such special cases.)

Reply:
So what you're saying is
```c++
uint8_t* buffer = new uint8_t[400];
Pixel* pixels = static_cast<Pixel*>(buffer);
```
is undefined but
```c++
Pixel* pixels = new Pixel[100];
uint8_t* buffer = static_cast<uint8_t*>(pixels);
```
is fine?

Reply:
> If you want to use the uint8_t* as an array, is it guaranteed that &bar.b == &bar.a + 1 (it is not, compilers are free to add whatever padding they wish).

This is why you static assert that sizeof(Pixel) == 4.

Reply:
We only need to assert sizeof(Pixel) == 4 once in the codebase to ensure this since what we're checking is compiler behavior, right? So a `static_assert(sizeof(Pixel) == 4)` after the declaration of `Pixel` would guarantee this?

(I'm currently reading about aliasing [here](https://developers.redhat.com/blog/2020/06/02/the-joys-and-perils-of-c-and-c-aliasing-part-1#objects_and_types_in_c_and_c__).)

Reply:
> So what you're saying is

No.

The language standard is saying that:

```c++
Pixel foo;

auto bar = reinterpret_cast<uint8_t&>(foo);
```

Is OK (because we're meeting a bunch of criteria, including it being the type of the first member). And that variants that are not equivalent, like accessing the second member, are not defined behaviour.

Reply:
I introduced the abstraction to help fix a bug I couldn't figure out, but if it's gonna cause this much trouble, let's shelve it. I got the bug figured out the old way.